File: leet/leet_628.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maximumProduct(vector<int>& nums) {
        int max_int = ~(1 << 31);
        int min_int = (1 << 31);
        int len = nums.size();
        if(len == 3) return nums[0] * nums[1] * nums[2];
        // max1 > max2 > max3 > ... > min2 > min1
        int max1 = min_int, max2 = min_int, max3 = min_int, min1 = max_int, min2 = max_int;
        // one pass: O(n)
        for(int i = 0; i < len; ++i)
        {
            if(nums[i] > max1) 
            { 
                // NOTE: reversely assign
                max3 = max2;
                max2 = max1;
                max1 = nums[i];
            }
            else if(nums[i] > max2) 
            {
                // NOTE: reversely assign
                max3 = max2;
                max2 = nums[i];
            }
            else if(nums[i] > max3)
            {
                max3 = nums[i];
            }
            // should be independent to calcu max
            if(nums[i] < min1)
            {
                min2 = min1;
                min1 = nums[i];
            }
            else if(nums[i] < min2)
            {
                min2 = nums[i];
            }
        }
        //cout << "max1 = " << max1 << endl;
        //cout << "max2 = " << max2 << endl;
        //cout << "max3 = " << max3 << endl;
        //cout << "min1 = " << min1 << endl;
        //cout << "min2 = " << min2 << endl;
        if(max1 <= 0) 
        {
            return max1 * max2 * max3;
        }
        else
        {
            if (min1 * min2 > max2 * max3) return min1 * min2 * max1;
            else return max1 * max2 * max3;
        }
    }
};

int main()
{
    int arr[] = {-4, -3, -2, 60};
    vector<int> vi(arr, arr+sizeof(arr)/sizeof(int));
    Solution solu;
    cout << solu.maximumProduct(vi) << endl;
    return 0;
}

File: leet/leet_628.py
#given an integer array, find 3 numbers whose product is maximum and output this maximum product
class Solution(object):
    def maximumProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_int = 2147483647
        min_int = -2147483648
        leng = len(nums)
        if leng == 3:
            return nums[0] * nums[1] * nums[2]
        max1 = min_int # 1st
        max2 = min_int # 2nd
        max3 = min_int # 3rd
        min1 = max_int # last
        min2 = max_int # 2nd last
        for i in range(leng):
            # calc the three max
            if nums[i] > max1:
                max3 = max2
                max2 = max1
                max1 = nums[i]
            elif nums[i] > max2:
                max3 = max2
                max2 = nums[i]
            elif nums[i] > max3:
                max3 = nums[i]
            # calc the two min: should be independent to calu max (min and max could be duplicate when leng <=5?)
            if nums[i] < min1:
                min2 = min1
                min1 = nums[i]
            elif nums[i] < min2:
                min2 = nums[i]
        # if all are less or equal than 0
        if max1 <= 0:
            return max1 * max2 * max3
        else:
            if min1 * min2 > max2 * max3:
                return min1 * min2 * max1
            else:
                return max1 * max2 * max3
if __name__ == '__main__':
    solu = Solution()
    print solu.maximumProduct([2, 1, 2, 3])

